walkthroughs/week-5-tdd/project/src/main/java/com/google/sps/FindMeetingQuery.java

```diff
@@ -15,9 +15,214 @@
+    Collection<String> optional = request.getOptionalAttendees();
+    
+    // sort the events to be in order by start time
+    List<Event> events_list = new ArrayList<Event>(events); 
+    Collections.sort(events_list, Event.ORDER_BY_START); 
+    List<TimeRange> possibleTimes = new ArrayList<>();
+
+    // if the duration of meeting is invalid, then there are no meeting times.
+    if (duration > TimeRange.getTimeInMinutes(23, 59) || duration < 0) {
+      return possibleTimes;
+    }
+    
+    // if there are no events for the day, then the meeting can be any time.
+    if (events.size() == 0) {
+      possibleTimes.add(TimeRange.WHOLE_DAY);
+      return possibleTimes;
+    }
+    
+    // If there are optional attendees
+    if (optional.size() > 0) {
+
+      List<String> allAttendees = new ArrayList<>(mandatory);
+      allAttendees.addAll(optional);
+      // Create new meeting request with mandatory and optional attendees as mandatory
+      MeetingRequest new_request = new MeetingRequest(allAttendees, duration);
+
+      // Find possible time ranges with all mandatory and all optional attendees
+      Collection<TimeRange> slotsOptional = intervalFinder(events_list, new_request);
+      
+      // if there is a time with all mandatory and optional attendees, then return
+      if (slotsOptional.size() > 0) {
+        return slotsOptional;
+      }
+      // if there is no time where everyone can make it, and there is more than one optional person
+      else if (optional.size() > 1) {
+        // find times for just the mandatory attendees
+        Collection<TimeRange> availableTimesMand = intervalFinder(events_list, request);
+
+        // store the times with the minimum amount of event overlaps
+        List<TimeRange> bestTimes = new ArrayList<>();
+        int minOverlaps = Integer.MAX_VALUE;
+
+        // containedSet will store the split ranges of the ranges of the availableTimesMand if an event that has an optional attendee is overlapping with the range for the availableTimesMand
+        Set<TimeRange> containedSet = new HashSet<>();
+        // Separate by optional attendee 
+        for (String opAttendee : optional) {
+          Collection<Event> eventsWithAttendee = eventFinderWithAttendee(events_list, opAttendee);
+          for (TimeRange range : availableTimesMand) {
+            containedSet.addAll(rangeSplitByNestedEvents(eventsWithAttendee, range, duration));
+          }
+        }
+        List<TimeRange> contained = new ArrayList<>(containedSet);
+
+        if (contained.size() > 0) {
+          for (TimeRange c : contained) {
+            // for every range in the contained, count the number of optional attendees that would not be able to make it
+            minOverlaps = optionalCounter(bestTimes, events, optional, c, minOverlaps);
+          }
+          return bestTimes;
+        }
+        // if there are no events that are contained within the mandatory ranges
```

Comment:
Might be easier to read if you include all the explanation of the logic in one comment beforehand rather than doing one comment at a time as we get deeper into the loop (i.e. move that second comment to the first to explain what the overall chunk of code is doing and why)